packages/ws-server/src/webSocketServer.ts

```diff
@@ -175,71 +176,84 @@ app.ws.use(async (ctx) => {
     methodsCounter.labels(method).inc();
     methodsCounterByIp.labels(ctx.request.ip, method).inc();
 
-    if (method === constants.METHODS.ETH_SUBSCRIBE) {
-      if (limiter.validateSubscriptionLimit(ctx)) {
-        const event = params[0];
-        const filters = params[1];
-        let subscriptionId;
-
-        if (event === constants.SUBSCRIBE_EVENTS.LOGS) {
-          try {
-            await validateSubscribeEthLogsParams(filters, requestIdPrefix);
-          } catch (error) {
-            logger.error(
-              error,
-              `${connectionIdPrefix} ${requestIdPrefix} Encountered error on ${
-                ctx.websocket.id
-              }, method: ${method}, params: ${JSON.stringify(params)}`,
-            );
-            response = jsonResp(request.id, error, undefined);
-            ctx.websocket.send(JSON.stringify(response));
-            return;
-          }
+    // Early return if subscription limit is exceeded
+    if (method === constants.METHODS.ETH_SUBSCRIBE && !limiter.validateSubscriptionLimit(ctx)) {
+      response = jsonResp(request.id, predefined.MAX_SUBSCRIPTIONS, undefined);
+      ctx.websocket.send(JSON.stringify(response));
+      return;
+    }
 
-          if (!getMultipleAddressesEnabled() && Array.isArray(filters.address) && filters.address.length > 1) {
-            response = jsonResp(
-              request.id,
-              predefined.INVALID_PARAMETER('filters.address', 'Only one contract address is allowed'),
-              undefined,
-            );
-          } else {
-            subscriptionId = relay.subs()?.subscribe(ctx.websocket, event, filters);
+    try {
+      switch (method) {
+        case constants.METHODS.ETH_SUBSCRIBE: {
+          const event = params[0];
+          const filters = params[1];
+          let subscriptionId;
+
+          switch (event) {
+            case constants.SUBSCRIBE_EVENTS.LOGS:
+              await validateSubscribeEthLogsParams(filters, requestIdPrefix);
+              if (!getMultipleAddressesEnabled() && Array.isArray(filters.address) && filters.address.length > 1) {
+                response = jsonResp(
+                  request.id,
+                  predefined.INVALID_PARAMETER('filters.address', 'Only one contract address is allowed'),
+                  undefined,
+                );
+              } else {
+                subscriptionId = relay.subs()?.subscribe(ctx.websocket, event, filters);
+              }
+              break;
+
+            case constants.SUBSCRIBE_EVENTS.NEW_HEADS:
+              if (process.env.WS_NEW_HEADS_ENABLED === 'true') {
+                ({ response, subscriptionId } = subscribeToNewHeads(
+                  filters,
+                  response,
+                  request,
+                  subscriptionId,
+                  ctx,
+                  event,
+                ));
+              }
```

Comment:
in case the feature is disabled, should we return an error response similar to `NEW_PENDING_TRANSACTIONS` such as `Unsupported method`?

this would need a test case.

Reply:
Done. Good catch.